File: backend/src/main/java/com/dreampath/service/rag/RagSearchService.java
package com.dreampath.service.rag;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;

@Service
@RequiredArgsConstructor
public class RagSearchService {

    @Value("${pinecone.api.key}")
    private String pineconeApiKey;

    @Value("${pinecone.host}")
    private String pineconeHost;

    private final OkHttpClient client = new OkHttpClient();

    public JSONArray search(float[] vector) {

        try {
            JSONObject json = new JSONObject();
            json.put("vector", vector);
            json.put("topK", 5);
            json.put("includeMetadata", true);

            RequestBody body = RequestBody.create(
                    json.toString(),
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url("https://" + pineconeHost + "/query")
                    .addHeader("Api-Key", pineconeApiKey)
                    .post(body)
                    .build();

            Response response = client.newCall(request).execute();
            String responseBody = response.body().string();

            JSONObject obj = new JSONObject(responseBody);
            return obj.getJSONArray("matches");

        } catch (Exception e) {
            throw new RuntimeException("Pinecone 검색 실패", e);
        }
    }
}

File: ai-service/services/rag/rag_search_service.py
from services.rag.pinecone_vector_service import PineconeVectorService
from services.vector.supabase_vector_repository import SupabaseVectorRepository
from pinecone import Pinecone


class RAGSearchService:

    def __init__(self):
        self.vector = PineconeVectorService()
        self.repo = SupabaseVectorRepository()
        self.pc = self.vector.pc
        self.index = self.vector.index

    def search(self, query_vector, top_k=5, filter=None):
        '''
        filter 예시: {"type": {"$eq": "job"}}
        '''
        response = self.index.query(
            vector=query_vector,
            top_k=top_k,
            filter=filter,
            include_metadata=True
        )
        return response.matches
